timeskew/timeskew.go

```diff
@@ -0,0 +1,117 @@
+package timeskew
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/beevik/ntp"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const (
+	// DefaultServer will be internally resolved to the closest available.
+	// also it rarely queries same server more than once.
+	DefaultServer = "pool.ntp.org"
+
+	// DefaultAttempts defines how many servers we will query
+	DefaultAttempts = 3
+
+	// DefaultUpdatePeriod defines how often time will be queried from ntp.
+	DefaultUpdatePeriod = 2 * time.Minute
+)
+
+type ntpQuery func(string) (*ntp.Response, error)
+
+func computeOffset(queryMethod ntpQuery, server string, attempts int) (time.Duration, error) {
+	offsets := make([]time.Duration, 5)
+	// lowest and highest will be cutoff to prevent reading data from out of sync ntp servers
+	for i := 0; i < attempts+2; i++ {
+		response, err := queryMethod(server)
+		if err != nil {
+			return 0, err
+		}
+		offsets[i] = response.ClockOffset
+	}
+	sort.SliceStable(offsets, func(i, j int) bool {
+		return offsets[i] > offsets[j]
+	})
+	var sum time.Duration
+	for i := 1; i <= attempts; i++ {
+		sum += offsets[i]
+	}
+	return sum / time.Duration(attempts), nil
```

Comment:
I would pick a median instead of an average. That way a broken (or spoofed) server won't be able to affect the value that much.
 
- avg(1,2,3,4) = 2,5;
- med(1,2,3,4) = 2,5;
---
- avg(-4,2,3,99) =25;
- med(-4,2,3,99) = 2,5;

Reply:
for that problem i am dropping min and max value, so for example if i get such 5 values:

-10, 1, 2, 3, 10

i will drop both -10 and 10, and only then compute a mean

Reply:
yeah, except if you already have a sorted slice it is enough to get the middle element out of it.
that looks more straightforward to me.
that logic with min/max removals slipped from my attention on the first read.